### DISCUSSION TITLE
Retrieve History of issue transitions

### DISCUSSION THREAD

**Discussion by zlogdanbr** [pycontribs/jira#1487] [Q&A] [answered]
Hi

I am trying to retrieve a field called History from JIRA. It describes the transitions that have been made for a specific issue. At my JIRA, it is displayed at the bottom of the issue where you can select to see Activity and then comments, worklog, etc. 

I am not sure if the JIRA lib supports it as it supports for example Worklog

```
issue.fields.worklog.worklogs                                 # list of Worklog objects
issue.fields.worklog.worklogs[0].author
issue.fields.worklog.worklogs[0].comment
issue.fields.worklog.worklogs[0].created
issue.fields.worklog.worklogs[0].id
issue.fields.worklog.worklogs[0].self
issue.fields.worklog.worklogs[0].started
issue.fields.worklog.worklogs[0].timeSpent
issue.fields.worklog.worklogs[0].timeSpentSeconds
issue.fields.worklog.worklogs[0].updateAuthor               
issue.fields.worklog.worklogs[0].updated
```

Worklog is empty in my case.  So, basically, what I would need is an object like

issue.fields.history which would allow me to iterate over the transitions made.

My user at JIRA is a regular user, we just csv file reports from JIRA and process them, but this field History contains these transitions we need to log .

Our JIRA admin told us we can directly access the field through 

https://jira.atlassian.com/browse/JRASERVER-27692 

Example:  https://jira.fis.dev/rest/api/2/issue/CMSUPPORT-141111?expand=changelog

which actually works for me if use the browser.  I actually think it would be difficult to do all the processing by hand, e.g using REST python lib to accomplish that. Mostly because of the authentication mechanism which I am struggling now.

If anyone has any idea about this, I would appreciate some help.

thanks
Daniel

**Comment by zlogdanbr** [comment, marked as answer]
I actually had to directly access the API using python request APIs.

this code will return a list of each transition

ID    TIMESTAMP(string)    Previous State   Current State

`
def rest_get_transitions_from_jira(url_issue, session, _debug):
    
    myjson = rest_get_jason_obj( url_issue, session )
    changelog_jason = myjson['changelog']        
    history = changelog_jason['histories'] 
    
    if _debug== True:
        rest_debug_jason( changelog_jason)

    transitions = []
      
    for history_item in history:
        
        the_item_in_history = history_item['items']
        id = history_item['id']
        timestamp = history_item['created']
                
        for _item in the_item_in_history:
            # will not add relevant time complexity because k is fixed  < 10
            for k in _item:               
                if k == 'field':
                    if _item[k] == 'status':
                        if _item['fromString'] != None:
                            transitions.append([ id, timestamp, _item['fromString'], _item['toString']])
                                                              
    return transitions
`
# session is session.get(url_jira, auth = HTTPBasicAuth(username, password))
# URL is https://<YOUR_JIRA>/rest/api/2/issue/<ISSUE>?expand=changelog
`def rest_get_jason_obj( url, session ):
    
    response = session.get( url )
    if response.status_code == 200:
        return response.json() 
    else:
        return None
 `

**Reply by maciejkoszykowski** [reply]
Hello @zlogdanbr 

There is a way to do it with jira python:
```
issue = jira.issue('<jira issue id>',expand='changelog')

changelog = issue.changelog

for history in changelog.histories:
    # print("history.created: " + history.created) # remove comment if you want to see that there are other history records    
    for item in history.items:    
        if item.field == 'status':            
            print(  "status changed on: " + history.created + 
                       " from status: " + item.fromString + "   to status: " + item.toString)             
```

Probably, you don't needed anymore, but maybe someone else is going to need it like I needed.
This way you avoid an extra work around with REST API calls. 

have a nice day, Maciej

**Comment by zlogdanbr** [comment]
HI  Maciej

Thanks a lot for answering! I actually handled the direct REST messages,
but this looks much better. I always prefer to use a library if there is
one.
I left the company I was working on this app, but I will forward it to the
person who is in charge now!

best regards
Daniel

On Mon, Mar 6, 2023 at 6:06 PM maciejkoszykowski ***@***.***>
wrote:

> Hello @zlogdanbr <https://github.com/zlogdanbr>
>
> There is a way to do it with jira python:
>
> issue = jira.issue('<jira issue id>',expand='changelog')
>
> changelog = issue.changelog
>
> for history in changelog.histories:
>     # print("history.created: " + history.created) # remove comment if you want to see that there are other history records
>     for item in history.items:
>         if item.field == 'status':
>             print(  "status changed on: " + history.created +
>                        " from status: " + item.fromString + "   to status: " + item.toString)
>
> Probably, you don't needed anymore, but maybe someone else is going to
> need it like I needed.
> This way you avoid an extra work around with REST API calls.
>
> have a nice day, Maciej
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pycontribs/jira/discussions/1487#discussioncomment-5220448>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANBG5WR2X5EHDLETSIRQW6LW2ZGTRANCNFSM6AAAAAAQCSLAYE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
“Debugging is twice as hard as writing the code in the first place.
Therefore, if you write the code as cleverly as possible, you are, by
definition, not smart enough to debug it.”  ( Brian Kernighan)
https://zlogdan.wordpress.com/
